### PR TITLE
refactor: consolidate experiment list selection state

### PR DESCRIPTION
**Pull request by ashtonG**
## Description

This refactors some items related to the experiment list selection ahead of fixing the bulk action apis:

1) selection state (selected/excluded experiment ids, selectAll state) is consolidated into a single object. Previously, selected experiment ids, excluded experiment ids, and the selectall state were stored separately. This is problematic -- because the selected experiment ids were the source of truth for which experiments were selected, the stored list would grow as the user paged through the results when in selectall mode. we now determine the final list of selected elements on the fly.

2) grid selection is now fully determined from the id selection. previously, the grid selection and id selection were maintained separately. this required a lot of code ensuring they stayed in sync. we now only derive the grid selection from the selected id list.

3) handleSelectionChange now no longer requires row ranges for `add-all` or `remove-all` change types. this allows for functions and effects that solely dealt with these change types to no longer rely on the experiments array.

4) project experiment list settings is no longer using useSettings. because the grid selection now only updates when the settings updates, the act of updating the selection felt slower than when using useSettings. we now directly interface with the settings store at the project level -- global settings still use useSettings.

5) new hook: useTypedParams. In order to keep the functionality of the `compare` parameter, we introduce a useTypedParams hook which allows typesafe parsing and setting of query params similar to the settings store api. this should allow us to finally fully migrate off of useSettings.

## Test Plan

Parameter/setttings behavior:
- navigate to a project experiment list with over 20 experiments in the list
- turn off infinite scroll and reload the page
- the experiment list should remain paginated
- click onto the second page
- the url should contain the query parameter `page=1`
- reload the page
- the experiment list should load the second page
- click back onto the first page
- the url should no longer contain the page query parameter
- toggle the comparison view on
- the url should contain the query parameter `compare=true`
- reload the page
- the experiment list should be in comparison mode
- remove the query parameter `compare=true` and reload the page
- the experiment list should remain in comparison mode
- add the query paramter `compare=false` and reload the page
- the experiment list should not be in comparison mode

Selection behavior:
- navigate to a project experiment list with over 20 experiments in the list
- select a single experiment in the list
- reload the page
- the experiment should remain selected
- navigate to the next page
- select a single experiment in the list
- navigate to the first page
- the selection status at the top of the table should show that one experiment is selected
- hold shift and select an experiment
- the experiment should be added to the selection
- in the selection column menu, choose select all
- the selection status at the top of the table should show that all experiments are selected
- deselect an experiment by clicking a checkbox
- the selection status at the top of the table should respond likewise

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
this is base work for WEB-1937

### CODE REVIEW

**Review by EmilyBonar — CHANGES_REQUESTED** [review]
I'm still working my way through the code, but I found some problems while running through the test plan.

- Going to the second page and reloading brings me back to first page
- Changing query param to `compare=false` didn’t work, it changed itself back to `true` and comparison view stayed open
- Holding shift while selecting an experiment didn’t seem to do anything; either way any selections off the page got replaced
- Similarly. if I select all experiments, deselect one (A), switch pages, and deselect another experiment (B), the first experiment I deselected (A) gets re-selected

**Comment by ashtonG** [comment]
@EmilyBonar just pushed a fix that should address selections across pages and the compare param not persisting properly. This should also mitigate one source of the page being reset. could you confirm when you have the time?

**Comment by EmilyBonar** [comment]
The compare view and deselection across pages (bullet point 4) bugs are fixed. The bug with the page number getting reset on refresh is still present.

I'm still not entirely sure what the shift behavior is supposed to be, but when I follow this section of the test plan:

> - the selection status at the top of the table should show that one experiment is selected
> - hold shift and select an experiment
> - the experiment should be added to the selection

The selection status shows that 2 experiments are selected, and the next experiment seems to be added to the selection whether or not shift is held.

**Comment by ashtonG** [comment]
updated the test plan -- i was thinking of the native glide table selection behavior, i've replaced it with how shift selections work currently. I've also pushed a potential fix for the page being reset by changing how we detect if the scroll took (previously we assumed that the table would be set up after a certain amount of renders with the glide table mounted, now we wait for the row in question to be in bounds) can you verify?

**Comment by EmilyBonar** [comment]
The page reset bug is fixed! The only step of the test plan I'm still having trouble with is

> the selection status at the top of the table should show that one experiment is selected

When I reach that step I see "2 of [total] experiments selected"

**Comment by ashtonG** [comment]
@EmilyBonar updated the test plans again -- looking at the main branch that behavior is acceptable.

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 1]
```diff
@@ -161,95 +172,114 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
...
-      handleError(formSetValidation.left, {
-        publicSubject: 'Unable to initialize filterset from settings',
+    let cleanup: () => void;
+    // we only want to init the filterset state from settings once and only when
+    // the settings have finished loading. subscribing to the project settings
+    // observable only works on initial loads -- the settings store will be
+    // fully loaded when navigating to the page in-router, so we check if we
+    // need to subscribe at all.
+    const initFormset = (ps: Loadable<ProjectSettings | null>) => {
+      ps.forEach((s) => {
```
Comment (EmilyBonar) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:189]: unrelated to this pr but `forEach` is a really misleading name for this method. something like `ifLoaded` seems like it would be more clear

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 2]
```diff
@@ -161,95 +172,114 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
...
+      initFormset(projectSettingsObs.get());
     } else {
-      const formset = formSetValidation.right;
-      formStore.init(formset);
+      cleanup = projectSettingsObs.subscribe((ps) => {
+        if (ps.isLoaded) {
+          cleanup();
+          initFormset(ps);
+        }
+      });
```
Comment (EmilyBonar) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:216]: nit: this part is hard to follow, a comment would be helpful.

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 3]
```diff
@@ -161,95 +172,114 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
...
+  // partition experiment list into not selected/selected with indices and experiments so we only iterate the result list once
+  const [excludedExperimentIds, selectedExperimentIds]: [
+    Map<number, { index: number; experiment: ExperimentItem }>,
+    Map<number, { index: number; experiment: ExperimentItem }>,
+  ] = useMemo(() => {
+    if (isLoadingSettings) {
+      return [new Map(), new Map()];
+    }
+    const selectedMap = new Map();
+    const excludedMap = new Map();
```
Comment (EmilyBonar) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:234]: super nit: you could define the selectedMap and excludedMap before the conditional and just return them if empty. saves you from constructing two additional maps

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 4]
```diff
@@ -161,95 +172,114 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
...
+    const selectedIdSet =
+      settings.selection.type === 'ONLY_IN' ? new Set(settings.selection.selections) : new Set();
```
Comment (EmilyBonar) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:236]: super nit: i think the following is slightly nicer? but up to you
```
const selectedIdSet = new Set(
    settings.selection.type === 'ONLY_IN' ? 
        settings.selection.selections : 
        []
    );
```

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx (hunk 3 above)
Comment (EmilyBonar) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:234]: Also it would be good to move the map types to here so that the rest of the code in the useMemo can benefit from the typechecking instead of having `Map<any, any>`

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx [hunk 5]
```diff
@@ -293,6 +316,36 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [isLoadingSettings]);
 
+  useEffect(() => {
```
Comment (EmilyBonar) [inline webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:319]: nit: this whole useEffect could use more commenting

**Review by EmilyBonar — APPROVED** [review]
Test plan passes and everything looks good! Just some code style comments

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx (hunk 1 above)
Comment (ashtonG) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:189]: i agree but see why we made the decision -- it's not unheard of to use part of the array api to describe the methods of a generic monad because lists are monads that developers are familiar with.

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx (hunk 5 above)
Comment (ashtonG) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:319]: will probably refactor this pattern to be clearer -- it's the same setup as the formset effect, but it's very verbose and we're doing a weird thing with the cleanup reference in order to get around `subscribe` only being called when the observable changes.

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx (hunk 3 above)
Comment (ashtonG) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:234]: good point!

File: webui/react/src/pages/F_ExpList/F_ExperimentList.tsx (hunk 5 above)
Comment (EmilyBonar) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:319]: Yeah, it's confusing in the other spot too, especially the cleanup

Comment (ashtonG) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:313]: @EmilyBonar refactored the effects -- is this easier to understand?

Comment (EmilyBonar) [reply webui/react/src/pages/F_ExpList/F_ExperimentList.tsx:313]: Yeah that's a lot better, thanks!
